libpromises/evalfunction.c

```diff
@@ -279,38 +279,49 @@ static JsonElement* VarRefValueToJson(const EvalContext *ctx, const FnCall *fp,
                     data);
             }
         }
-        default:
-            *allocated = true;
 
+        default:              /* var not found, lets look for var[whatever] */
+
+            *allocated = true;
             {
                 VariableTableIterator *iter = EvalContextVariableTableFromRefIteratorNew(ctx, ref);
                 convert = JsonObjectCreate(10);
                 const size_t ref_num_indices = ref->num_indices;
                 char *last_key = NULL;
                 Variable *var;
 
+                /* Iterate over all *array* variables with that base name. */
                 while ((var = VariableTableIteratorNext(iter)) != NULL)
                 {
                     JsonElement *holder = convert;
+                    /* Search for arrays with more indices than var. */
                     if (var->ref->num_indices - ref_num_indices == 1)
                     {
+                        /* Just one extra index. */
                         last_key = var->ref->indices[ref_num_indices];
                     }
                     else if (var->ref->num_indices - ref_num_indices > 1)
                     {
+                        /* More than one indices. */
                         Log(LOG_LEVEL_DEBUG, "%s: got ref with starting depth %zu and index count %zu",
                             fp_name, ref_num_indices, var->ref->num_indices);
                         for (int index = ref_num_indices; index < var->ref->num_indices-1; index++)
                         {
                             JsonElement *local = JsonObjectGet(holder, var->ref->indices[index]);
                             if (local == NULL)
                             {
+                                /* TODO comment WHY? */
```

Comment:
If the holder has nothing under that key, we create a new object to go under that key.